examples_apps/run_gate_synthesis.py

```diff
@@ -0,0 +1,469 @@
+
+eng = sf.Engine('tf', backend_options={"cutoff_dim": cutoff, "batch_size": gate_cutoff})
+
+######################################################################
+# Here, we use the ``batch_size`` argument to perform the optimization in
+# parallel - each batch calculates the variational quantum circuit acting
+# on a different input Fock state: :math:`U(\vec{\theta}) | n\rangle`.
+#
+
+in_state = np.arange(gate_cutoff)
+
+# extract action of the target unitary acting on
+# the allowed input fock states.
+target_kets = np.array([target_unitary[:, i] for i in in_state])
+target_kets = tf.constant(target_kets, dtype=tf.complex64)
+
+######################################################################
+# Using this target unitary, we define the cost function we would like to
+# minimize. We must use TensorFlow functions to manipulate this data, as well
+# as a ``GradientTape`` to keep track of the corresponding gradients!
+#
+def cost(weights):
+    # Create a dictionary mapping from the names of the Strawberry Fields
+    # free parameters to the TensorFlow weight values.
+    mapping = {p.name: w for p, w in zip(sf_params.flatten(), tf.reshape(weights, [-1]))}
+
+    # Run engine
+    state = eng.run(prog, args=mapping).state
+
+    # Extract the statevector
+    ket = state.ket()
+
+    # overlaps
+    overlaps = tf.math.real(tf.einsum('bi,bi->b', tf.math.conj(target_kets), ket))
+    mean_overlap = tf.reduce_mean(overlaps)
+
+    # Objective function to minimize
+    cost = tf.abs(tf.reduce_sum(overlaps - 1))
+    return cost, overlaps, ket
+
+
+
+######################################################################
+# Now that the cost function is defined, we can define and run the
+# optimization. Below, we choose the Adam optimizer that is built into
+# TensorFlow.
+#
+
+# Using Adam algorithm for optimization
+opt = tf.keras.optimizers.Adam(learning_rate=lr)
+
+
+######################################################################
+# We then loop over all repetitions, storing the best predicted overlap
+# value.
+#
+
+overlap_progress = []
+cost_progress = []
+
+# Run optimization
+for i in range(reps):
+
+    # reset the engine if it has already been executed
+    if eng.run_progs:
+        eng.reset()
+
+    # one repetition of the optimization
+    with tf.GradientTape() as tape:
+        loss, overlaps_val, ket_val = cost(weights)
+
+    # calculate the mean overlap
+    # This gives us an idea of how the optimization is progressing
+    mean_overlap_val = np.mean(overlaps_val)
+
+    # store cost at each step
+    cost_progress.append(loss)
+    overlap_progress.append(overlaps_val)
+
+    # one repetition of the optimization
+    gradients = tape.gradient(loss, weights)
+    opt.apply_gradients(zip([gradients], [weights]))
+
+    # Prints progress at every 100 reps
+    if i % 1 == 0:
+        # print progress
+        print("Rep: {} Cost: {:.4f} Mean overlap: {:.4f}".format(i, loss, mean_overlap_val))
```

Comment:
Note that after 300 steps the I got a mean overlap of 90% - whereas in the paper we were looking at 99%+ and in the previous Jupyter notebook version we got to 98%+.

I think this is due to the number of iterations dropping to 300 from 1000 and the depth also going from 25 to 20. Is the motivation to speed things up? I guess it's ok, perhaps we could add a note saying that higher mean overlaps are possible if you keep optimizing?

Reply:
Thanks for that @trbromley! Yes, the motivation was speeding up, but these are fair points. Added a note on this for increasing the number of reps at the end of the tutorial. Also added certain updates and now training seems to be faster for me locally.

Reply:
Yep, this is just to reduce the time taken. We have two options:

* Leave it as is, and at the bottom of the tutorial, have a note box letting users know they can download the tutorial, and increase the number of steps for even better results!

* Turn it into a non-executable tutorial, and hardcode in the results.

As we're about to move the tutorials out to their own repository, I'm somewhat leaning towards option 1. Once it's moved over to it's own repo and building on CircleCI, we can just increase the number of reps.